### ISSUE TITLE
[Clang] Deduction guide wrongly produces non-deducible template parameter error when it can be deduced for alias template

### ISSUE THREAD

**Issue by ednolan** [llvm/llvm-project#126489]
Given the following code:
```cpp
#include <type_traits>

template<typename T, typename V>
struct Base {
    template<typename U>
    Base(U);
};

template<typename T, typename V>
Base(V) -> Base<T, V*>;

template<typename V>
using Int = Base<int, V*>;

Int x(10.0f);
static_assert(std::is_same_v<decltype(x), Base<int, float*>>);
```

Clang trunk produces the following error:
```
<source>:10:1: error: deduction guide template contains a template parameter that cannot be deduced
   10 | Base(V) -> Base<T, V*>;
      | ^
<source>:9:19: note: non-deducible template parameter 'T'
    9 | template<typename T, typename V>
      |                   ^
```

But the same code is accepted by GCC, MSVC, and EDG.

Compiler Explorer link: https://godbolt.org/z/MaE9oh837

Before P1814, it was accurate that the deduction guide Clang is complaining about did contain a non-deducible template parameter, but this is no longer the case now that the template parameter can be deduced from the alias template.

Moreover, the Clang AST in Compiler Explorer shows that the desired deduction guide is actually being generated:

```
<deduction guide for Int> 'auto (V) -> Base<int, V *>'
```

This appears to be a rejects-valid bug from Clang that impedes usage of deduction guides for alias templates.

**Comment by cor3ntin** [comment]
@hokein

**Comment by ednolan** [comment]
@mizvekov

**Comment by mizvekov** [comment]
@antangelo

**Comment by agraham-roku** [comment]
Related or dupe of [#111216](https://github.com/llvm/llvm-project/issues/111216)

**Comment by term-est** [comment]
Okay, so Clang is compliant here.

[temp.param]/19 says:

A template parameter of a deduction guide template that does not have a default argument shall be deducible from the parameter-type-list of the deduction guide template.

So,
```c++
template<typename T, typename V>
Base(V) -> Base<T, V*>;
```
is ill-formed: T isn't deducible from the parameter-type-list. [temp.res.general] permits the validity of the template to be checked prior to instantiation, so subsequent alias-template deduction-guide formation cannot rescue the declaration.

P1814 does not relax this requirement or otherwise make this usage well-formed. This therefore doesn't appear to be a Clang bug or a Core wording defect; making it well-formed would be a design change. I'll pursue that in a paper instead.

**Comment by zygoloid** [comment]
> P1814 does not relax this requirement or otherwise make this usage well-formed. This therefore doesn't appear to be a Clang bug or a Core wording defect; making it well-formed would be a design change. I'll pursue that in a paper instead.

CWG might consider handling it as a core issue nonetheless; it's worth asking, since the core issue process is lighter-weight than the paper process.
